### ISSUE TITLE
RET506 doesn't work for raise in try/except block

### ISSUE THREAD

**Issue by Greesb** [astral-sh/ruff#9625]
Hi,

I found that the following code does not raise an error with the `RET506` rule enabled (or any of the other rules in `RET`):
```python
def test() -> str:
    try:
        print("test")
    except Exception:
        raise
    else:
        return "test"
```

Since the `except` block is always raising an error, shouldn't the `RET506` rule show an error here ?

Thanks and keep up the great work on this wonderful tool !

**Comment by dhruvmanila** [comment]
Good point. Yes, I think it should be raised here atleast for your first example of `RET505`. I guess you meant `raise` on the last line in `test_ret506` function? If so, then yes for that as well.

We need to make sure that there isn't a `finally` block here because then the execution order will be different. It's `try`, `else`, `finally` and removing `else` would alter the order.

**Comment by Greesb** [comment]
> I guess you meant `raise` on the last line in `test_ret506` function? If so, then yes for that as well.

Whether it be a `return` or a `raise` in the `else` for both case it shouldn't matter, right ? In both case it won't be reached because of the `raise` / `return` in the `except` clause.

**Comment by trag1c** [comment]
Yep, it's all about the `except` body. I think you should've commented the `else:` line since that's the violating one 👍 (which could be why Dhruv thought the `raise`/`return` in the `else` body matters)

**Comment by Greesb** [comment]
> Yep, it's all about the `except` body. I think you should've commented the `else:` line since that's the violating one 👍 (which could be why Dhruv thought the `raise`/`return` in the `else` body matters)

Indeed, edited it as you said for better clarity.
